winlogbeat/eventlog/wineventlog.go

```diff
@@ -432,6 +426,47 @@ func (l *winEventLog) buildRecordFromXML(x []byte, recoveredErr error) Record {
 	return r
 }
 
+// winMetaCache retrieves and caches WinMeta tables by provider name.
+// It is a cut down version of the PublisherMetadataStore caching in wineventlog.Renderer.
+type winMetaCache struct {
+	mu     sync.RWMutex
+	cache  map[string]*winevent.WinMeta
+	logger *logp.Logger
+}
+
+func newWinMetaCache() winMetaCache {
+	return winMetaCache{cache: make(map[string]*winevent.WinMeta), logger: logp.L()}
+}
+
+func (c *winMetaCache) winMeta(provider string) *winevent.WinMeta {
+	c.mu.RLock()
+	m, ok := c.cache[provider]
+	c.mu.RUnlock()
+	if ok {
+		return m
+	}
+
+	// Upgrade lock.
+	defer c.mu.Unlock()
+	c.mu.Lock()
+
+	// Did the cache get updated during lock upgrade?
+	if m, ok := c.cache[provider]; ok {
+		return m
+	}
+
+	s, err := win.NewPublisherMetadataStore(win.NilHandle, provider, c.logger)
```

Comment:
I think `s.Close()` needs to be called to free the publisher metadata handle.

Reply:
This should be back-ported to 7.17 then I think, otherwise we have a resource leak on that version.

Done.